README.md
# opentracing-metrics-tracer

[![Build Status](https://travis-ci.org/RisingStack/opentracing-metrics-tracer.svg?branch=master)](https://travis-ci.org/RisingStack/opentracing-metrics-tracer)  

Exports cross-process metrics via OpenTracing instrumentation to reporters: Prometheus.  
It's capable to measure operation characteristics in a distributed system like microservices.  

It also makes possible to reverse engineer the infrastructure topology as we know the initiators

## Available Reporters

- [Prometheus](https://prometheus.io/) via [prom-client](https://github.com/siimon/prom-client)

## Getting started

```js
const MetricsTracer = require('@risingstack/opentracing-metrics-tracer')
const prometheusReporter = new MetricsTracer.PrometheusReporter()
const metricsTracer = new MetricsTracer('my-service', [prometheusReporter])

// Instrument
const span = metricsTracer.startSpan('my-operation')
span.finish()
...

app.get('/metrics', (req, res) => {
  res.set('Content-Type', MetricsTracer.PrometheusReporter.Prometheus.register.contentType)
  res.end(prometheusReporter.metrics())
})
```

### With auto instrumentation and multiple tracers

Check out: https://github.com/RisingStack/opentracing-auto

```js
// Prometheus metrics tracer
const MetricsTracer = require('@risingstack/opentracing-metrics-tracer')
const prometheusReporter = new MetricsTracer.PrometheusReporter()
const metricsTracer = new MetricsTracer('my-service', [prometheusReporter])

// Jaeger tracer (classic distributed tracing)
const jaeger = require('jaeger-client')
const UDPSender = require('jaeger-client/dist/src/reporters/udp_sender').default
const sampler = new jaeger.RateLimitingSampler(1)
const reporter = new jaeger.RemoteReporter(new UDPSender())
const jaegerTracer = new jaeger.Tracer('my-server-pg', reporter, sampler)

// Auto instrumentation
const Instrument = require('@risingstack/opentracing-auto')
const instrument = new Instrument({
  tracers: [metricsTracer, jaegerTracer]
})

// Rest of your code
const express = require('express')
const app = express()

app.get('/metrics', (req, res) => {
  res.set('Content-Type', MetricsTracer.PrometheusReporter.Prometheus.register.contentType)
  res.end(prometheusReporter.metrics())
})
```

### Example

See [example server](/example/server.js).

```sh
node example/server
curl http://localhost:3000
curl http://localhost:3000/metrics
```

## API

`const Tracer = require('@risingstack/opentracing-metrics-tracer')`

### new Tracer(serviceKey, [reporter1, reporter2, ...])

- **serviceKey** *String*, *required*, unique key that identifies a specific type of service *(for example: my-frontend-api)*
- **reporters** *Array of reporters*, *optional*, *default:* []

[OpenTracing](https://github.com/opentracing/opentracing-javascript) compatible tracer, for the complete API check out the official [documentation](https://opentracing-javascript.surge.sh/).

### new Tracer.PrometheusReporter([opts])

- **opts** *Object*, *optional*
- **opts.ignoreTags** *Object*, *optional*
  - Example: `{ ignoreTags: { [Tags.HTTP_URL]: /\/metrics$/ } }` to ignore Prometheus scraper

Creates a new Prometheus reporter.

### Tracer.PrometheusReporter.Prometheus

Exposed [prom-client](https://github.com/siimon/prom-client).

## Reporters

### Prometheus Reporter

Exposes metrics in Prometheus format via [prom-client](https://github.com/siimon/prom-client)

#### Metrics

- [operation_duration_seconds](#operation_duration_seconds)
- [http_request_duration_seconds](#http_request_duration_seconds)

##### operation_duration_seconds

Always measured.  
Sample output: Two distributed services communicate over the network.

```
# HELP operation_duration_seconds Duration of operations in second
# TYPE operation_duration_seconds histogram
operation_duration_seconds_bucket{le="0.005",parent_service="my-parent-service",name="my-operation" 0
operation_duration_seconds_bucket{le="0.01",parent_service="my-parent-service",name="my-operation" 0
operation_duration_seconds_bucket{le="0.025",parent_service="my-parent-service",name="my-operation" 0
operation_duration_seconds_bucket{le="0.05",parent_service="my-parent-service",name="my-operation" 0
operation_duration_seconds_bucket{le="0.1",parent_service="my-parent-service",name="my-operation" 1
operation_duration_seconds_bucket{le="0.25",parent_service="my-parent-service",name="my-operation" 1
operation_duration_seconds_bucket{le="0.5",parent_service="my-parent-service",name="my-operation" 2
operation_duration_seconds_bucket{le="1",parent_service="my-parent-service",name="my-operation" 2
operation_duration_seconds_bucket{le="2.5",parent_service="my-parent-service",name="my-operation" 2
operation_duration_seconds_bucket{le="5",parent_service="my-parent-service",name="my-operation" 2
operation_duration_seconds_bucket{le="10",parent_service="my-parent-service",name="my-operation" 2
operation_duration_seconds_bucket{le="+Inf",parent_service="my-parent-service",name="my-operation" 2
operation_duration_seconds_sum{parent_service="my-parent-service",name="my-operation" 0.4
operation_duration_seconds_count{parent_service="my-parent-service",name="my-operation" 2
```

##### http_request_duration_seconds

Measured only when the span is tagged with `SPAN_KIND_RPC_SERVER` and any of `HTTP_URL`, `HTTP_METHOD` or `HTTP_STATUS_CODE`.  
Sample output:
```
# HELP http_request_handler_duration_seconds Duration of HTTP requests in second
# TYPE http_request_handler_duration_seconds histogram
http_request_handler_duration_seconds_bucket{le="0.005",parent_service="my-parent-service",method="GET",code="200",name="http_request" 0
http_request_handler_duration_seconds_bucket{le="0.01",parent_service="my-parent-service",method="GET",code="200",name="http_request" 0
http_request_handler_duration_seconds_bucket{le="0.025",parent_service="my-parent-service",method="GET",code="200",name="http_request" 0
http_request_handler_duration_seconds_bucket{le="0.05",parent_service="my-parent-service",method="GET",code="200",name="http_request" 0
http_request_handler_duration_seconds_bucket{le="0.1",parent_service="my-parent-service",method="GET",code="200",name="http_request" 1
http_request_handler_duration_seconds_bucket{le="0.25",parent_service="my-parent-service",method="GET",code="200",name="http_request" 1
http_request_handler_duration_seconds_bucket{le="0.5",parent_service="my-parent-service",method="GET",code="200",name="http_request" 2
http_request_handler_duration_seconds_bucket{le="1",parent_service="my-parent-service",method="GET",code="200",name="http_request" 2
http_request_handler_duration_seconds_bucket{le="2.5",parent_service="my-parent-service",method="GET",code="200",name="http_request" 2
http_request_handler_duration_seconds_bucket{le="5",parent_service="my-parent-service",method="GET",code="200",name="http_request" 2
http_request_handler_duration_seconds_bucket{le="10",parent_service="my-parent-service",method="GET",code="200",name="http_request" 2
http_request_handler_duration_seconds_bucket{le="+Inf",parent_service="my-parent-service",method="GET",code="200",name="http_request" 2
http_request_handler_duration_seconds_sum{parent_service="my-parent-service",method="GET",code="200",name="http_request" 0.4
http_request_handler_duration_seconds_count{parent_service="my-parent-service",method="GET",code="200",name="http_request" 2
```

## Future and ideas

This library is new, in the future we could measure much more useful and specific metrics with it.  
Please share your ideas in a form of issues or pull-requests.


CHANGELOG.md
<a name="2.1.0"></a>
## 2.1.0 (2017-09-29)


#### Features

* **tracer:** add ignoreTags option ([acd7c58a](git+https://github.com/RisingStack/opentracing-metrics-tracer.git/commit/acd7c58a))


<a name="2.0.1"></a>
### 2.0.1 (2017-09-04)


#### Bug Fixes

* **reporter:** report empty parent service key as "unknown" ([22a89a58](git+https://github.com/RisingStack/opentracing-metrics-tracer.git/commit/22a89a58))


<a name="2.0.0"></a>
## 2.0.0 (2017-09-04)


#### Bug Fixes

* **reporter:** rename http_request_duration_seconds_bucket to http_request_handler_duration_sec ([b07b5212](git+https://github.com/RisingStack/opentracing-metrics-tracer.git/commit/b07b5212))


#### Breaking Changes

* metrics renamed http_request_duration_seconds_bucket to http_request_handler_duration_seconds_bucket

 ([b07b5212](git+https://github.com/RisingStack/opentracing-metrics-tracer.git/commit/b07b5212))


<a name="1.3.0"></a>
## 1.3.0 (2017-09-04)


#### Features

* **example:** add child operation ([7a260dfa](git+https://github.com/RisingStack/opentracing-metrics-tracer.git/commit/7a260dfa))
* **reporter:** report parent_service with http_request_duration_seconds ([a06e2bc0](git+https://github.com/RisingStack/opentracing-metrics-tracer.git/commit/a06e2bc0))


<a name="1.2.0"></a>
## 1.2.0 (2017-09-03)


<a name="1.1.0"></a>
## 1.1.0 (2017-09-03)


#### Features

* **tracer:** public getters for Span and SpanContext ([a5b54c15](git+https://github.com/RisingStack/opentracing-metrics-tracer.git/commit/a5b54c15))


<a name="1.0.0"></a>
## 1.0.0 (2017-09-02)


#### Features

* **reporter:** add Prometheus reporter ([74c2b189](git+https://github.com/RisingStack/opentracing-metrics-tracer.git/commit/74c2b189))
* **tracer:** add Reference, Span, SpanContext and Tracer models ([e87df8e5](git+https://github.com/RisingStack/opentracing-metrics-tracer.git/commit/e87df8e5))



src/reporters/PrometheusReporter.js
'use strict'

const assert = require('assert')
const Prometheus = require('prom-client')
const { Tags } = require('opentracing')
const Span = require('../tracer/Span')

const DURATION_HISTOGRAM_BUCKETS = [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10]
const METRICS_NAME_OPERATION_DURATION_SECONDS = 'operation_duration_seconds'
const METRICS_NAME_HTTP_REQUEST_HANDLER_DURATION_SECONDS = 'http_request_handler_duration_seconds'
const LABEL_PARENT_SERVICE_UNKNOWN = 'unknown'

/**
* Observe span events and expose them in Prometheus metrics format
* @class PrometheusReporter
*/
class PrometheusReporter {
  /**
  * @static getParentServiceKey
  * @param {Span} span
  * @return {String} parentService
  */
  static getParentService (span) {
    const spanContext = span.context()
    const parentService = spanContext.parentServiceKey() || LABEL_PARENT_SERVICE_UNKNOWN

    return parentService
  }

  /**
  * @constructor
  * @param {Object} [options={}]
  * @param {Object} [options.ignoreTags={}]
  * @returns {PrometheusReporter}
  */
  constructor ({ ignoreTags = {} } = {}) {
    this._registry = new Prometheus.Registry()
    this._options = {
      ignoreTags
    }

    // Initialize metrics
    this._metricsOperationDurationSeconds()
  }

  /**
  * Returns with the reporter's metrics in Prometheus format
  * @method metrics
  * @returns {Object} metrics
  */
  metrics () {
    return this._registry.metrics()
  }

  /**
  * Called by Tracer when a span is finished
  * @method reportFinish
  * @param {Span} span
  */
  reportFinish (span) {
    assert(span instanceof Span, 'span is required')

    // Ignore by tag value
    const isIgnored = Object.entries(this._options.ignoreTags).some(([tagKey, regexp]) => {
      const tagValue = span.getTag(tagKey)
      return tagValue && tagValue.match(regexp)
    })

    if (isIgnored) {
      return
    }

    // Operation metrics
    this._reportOperationFinish(span)

    // HTTP Request
    if (span.getTag(Tags.SPAN_KIND_RPC_SERVER) &&
      (span.getTag(Tags.HTTP_URL) || span.getTag(Tags.HTTP_METHOD) || span.getTag(Tags.HTTP_STATUS_CODE))) {
      this._reportHttpRequestFinish(span)
    }
  }

  /**
  * Observe operation metrics
  * @method _reportOperationFinish
  * @private
  * @param {Span} span
  */
  _reportOperationFinish (span) {
    assert(span instanceof Span, 'span is required')

    this._metricsOperationDurationSeconds()
      .labels(PrometheusReporter.getParentService(span), span.operationName())
      .observe(span.duration() / 1000)
  }

  /**
  * Observe HTTP request metrics
  * @method _reportHttpRequestFinish
  * @private
  * @param {Span} span
  */
  _reportHttpRequestFinish (span) {
    assert(span instanceof Span, 'span is required')

    this._metricshttpRequestDurationSeconds()
      .labels(
        PrometheusReporter.getParentService(span),
        span.getTag(Tags.HTTP_METHOD),
        span.getTag(Tags.HTTP_STATUS_CODE)
      )
      .observe(span.duration() / 1000)
  }

  /**
  * Singleton to get operation duration metrics
  * @method _metricsOperationDurationSeconds
  * @private
  * @return {Prometheus.Histogram} operationDurationSeconds
  */
  _metricsOperationDurationSeconds () {
    let operationDurationSeconds = this._registry.getSingleMetric(METRICS_NAME_OPERATION_DURATION_SECONDS)

    if (!operationDurationSeconds) {
      operationDurationSeconds = new Prometheus.Histogram({
        name: METRICS_NAME_OPERATION_DURATION_SECONDS,
        help: 'Duration of operations in second',
        labelNames: ['parent_service', 'name'],
        buckets: DURATION_HISTOGRAM_BUCKETS,
        registers: [this._registry]
      })
    }

    return operationDurationSeconds
  }

  /**
  * Singleton to get HTTP request duration metrics
  * @method _metricshttpRequestDurationSeconds
  * @private
  * @return {Prometheus.Histogram} httpRequestDurationSeconds
  */
  _metricshttpRequestDurationSeconds () {
    let httpRequestDurationSeconds = this._registry.getSingleMetric(METRICS_NAME_HTTP_REQUEST_HANDLER_DURATION_SECONDS)

    if (!httpRequestDurationSeconds) {
      httpRequestDurationSeconds = new Prometheus.Histogram({
        name: METRICS_NAME_HTTP_REQUEST_HANDLER_DURATION_SECONDS,
        help: 'Duration of HTTP requests in second',
        labelNames: ['parent_service', 'method', 'code'],
        buckets: DURATION_HISTOGRAM_BUCKETS,
        registers: [this._registry]
      })
    }

    return httpRequestDurationSeconds
  }
}

PrometheusReporter.Prometheus = Prometheus
PrometheusReporter.LABEL_PARENT_SERVICE_UNKNOWN = LABEL_PARENT_SERVICE_UNKNOWN

module.exports = PrometheusReporter
